========
CLIP-PyL
========

The is the ALPHA release of the CLIP-PyL package, which contains scripts that can parse "crosslink signatures" from aligned CLIP-seq data.

Currently, HITS-CLIP is the only CLIP-seq method that is supported. However, PAR-CLIP and iCLIP data will be supported if considerable interest arises.

This tool requires that the user pre-processes the "raw" HITS-CLIP reads by clipping the adapter sequences. I recommend the FASTQ/A Clipper tool that is contained with the [FASTX toolkit](hannonlab.cshl.edu/fastx_toolkit/) for this task. However, other options exist. The user must also align the preprocessed reads to a reference genome assembly. Currently, single-end read alignments generated by the [bwa](bio-bwa.sourceforge.net/bwa.shtml) fast read alignment software are supported. Finally, the user must use the [SAMTools software](http://www.htslib.org/) to generate indexed bam files. CLIP-PyL takes the indexed bam files and returns a set of bedgraph files containing basewise crosslink metrics. The crosslink metrics that are generated include single base-pair deletions, fragment termini and raw coverage. Note that the bedgraph files can be viewed with most genome browser tools (e.g. [Broad's Integrated Genome Viewer](http://www.broadinstitute.org/igv/) or [UCSC's Genome Browser](http://genome.ucsc.edu/)).

The CLIP-PyL package also contains utilities to produce pdf files containing coverage map graphics for sets of genome intervals that are specified by the user in the form of a bed file. Cis-element regions can also be indicated in the output graphics if the user supplies additional bed files containing the locations of these elements. CLIP-PyL utilizies the thrid-party graphical back-end known as [matplotlib](http://matplotlib.org/) to generate these coverage graphics.

    #!/usr/bin/env python3
    
    from towelstuff import location
    from towelstuff import utils
    
    if utils.has_towel():
        print "Your towel is located:", location.where_is_my_towel()

Installation
============

Dependencies
------------

* python >= 3.2 https://www.python.org/

* pysam https://pypi.python.org/pypi/pysam

* matplotlib http://matplotlib.org/

